/*
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        https://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

syntax = "proto3";
import "SerializableDataType.proto";
import "SerializableIdentifier.proto";
package NES;

message SerializableField {
  SerializableIdentifier name = 1;
  SerializableDataType type = 2;
  uint64 operator_id = 3;
}

message SerializableUnboundField {
  SerializableIdentifier name = 1;
  SerializableDataType type = 2;
}

message SerializableUnboundSchema {
  repeated SerializableUnboundField fields = 1;
}

message SerializableSchema {

  enum MemoryLayoutType {
    ROW_LAYOUT = 0;
    COL_LAYOUT = 1;
  }

  repeated SerializableField fields = 1;
  map<uint64, SerializableIdentifierList> aliases = 2;
  MemoryLayoutType layoutType = 3;
}

